Sources/ApolloCore/CodegenCore/SelectionSet.swift

```diff
@@ -0,0 +1,67 @@
+public enum SelectionSetType<S: GraphQLSchema> {
+  case ObjectType(S.ObjectType)
+  case Interface(S.Interface)
+  case Union(S.Union)
+}
+
+public protocol SelectionSet: ResponseObject, Equatable {
+
+  associatedtype Schema: GraphQLSchema
+
+  /// The GraphQL type for the `SelectionSet`.
+  ///
+  /// This may be a concrete type (`ConcreteType`) or an abstract type (`Interface`).
+  static var __parentType: SelectionSetType<Schema> { get }
+}
+
+extension SelectionSet {
+
+  var __objectType: Schema.ObjectType { Schema.ObjectType(rawValue: __typename) ?? .unknownCase }
+
+  var __typename: String { data["__typename"] }
+
+  /// Verifies if a `SelectionSet` may be converted to a different `SelectionSet` and performs
+  /// the conversion.
+  ///
+  /// - Warning: This function is not supported for use outside of generated call sites.
+  /// Generated call sites are guaranteed by the GraphQL compiler to be safe.
+  /// Unsupported usage may result in unintended consequences including crashes.
+  func _asType<T: SelectionSet>() -> T? where T.Schema == Schema {
+    guard case let __objectType = __objectType, __objectType != .unknownCase else { return nil }
```

Comment:
can we not do this as a one-liner please?

Reply:
Sure, I'll clean that up as I start working on implementing the codegen. These aren't even tested yet (they will be). I just stubbed them in to make the development easier as I start writing the codegen.